hardware/is_camera/src/camera.cc

```diff
@@ -306,19 +337,66 @@ namespace is_camera {
         bayer_img_msg_buffer_[i]->step = kImageWidth;
         bayer_img_msg_buffer_[i]->data.resize(kImageWidth * kImageHeight);
       }
-  }
-  // Enable or disable the feature timer
-  bool CameraNodelet::EnableBayerService(ff_msgs::SetBool::Request & req,
-                     ff_msgs::SetBool::Response & res) {
-    if (req.enable && !bayer_enable_) {
-      EnableBayer();
-    } else if (!req.enable && bayer_enable_) {
+    } else if (!enable && bayer_enable_) {
       bayer_pub_.shutdown();
+    } else {
+      return;
+    }
+    enable = bayer_enable_;
+  }
+
+  // Timer that periodically changes camera exposure based on the captured image's histogram
+  void CameraNodelet::AutoExposure() {
+    // If images not being captured, do not adjust exposure
+    if (!thread_running_ || ((pub_.getNumSubscribers() == 0) && (getNumBayerSubscribers() == 0))) return;
+
+    // Get last generated image from buffer (index incremented after image completed)
+    sensor_msgs::ImagePtr& grey_image = img_msg_buffer_[(img_msg_buffer_idx_ - 1) % kImageMsgBuffer];
+
+    cv::Mat input(kImageHeight, kImageWidth,
+                     cv::DataType<uint8_t>::type,
+                     &(grey_image->data[0]),
+                     kImageWidth);
+
+    // Calculate Histogram
+    cv::Mat hist;
+    int histSize = 5;
+    float range[] = {0, 256};  // the upper boundary is exclusive
+    const float* histRange[] = { range };
+    bool uniform = true, accumulate = false;
+    cv::calcHist(&input, 1, 0, cv::Mat(), hist, 1, &histSize, histRange, uniform, accumulate);
+
+    // Calculate mean sample value
+    double mean_sample_value = 0;
+
+    for (int i = 0; i < histSize; ++i) {
+      mean_sample_value += hist.at<float>(i) * (i+1);
+    }
+    mean_sample_value /= (kImageHeight * kImageWidth);
+
+    // Control Auto exposure with a PI controller
+    err_p_ = desired_msv_ - mean_sample_value;
+
+    // Don't change exposure if we're close enough. Changing too often slows
+    // down the data rate of the camera.
+    if (std::abs(err_p_) > 0.2) {
+      // Calculate PI coefficients
+      err_i_ += err_p_;
+      if (std::abs(err_i_) > max_i_) {
+        err_i_ = (err_i_ > 0) ? max_i_ : - max_i_;
+      }
+      // Update camera exposure parameter
+      camera_auto_exposure_ = std::round(camera_exposure_ + k_p_ * err_p_ + k_i_ * err_i_);
+      if (camera_auto_exposure_ < 0) camera_auto_exposure_ = 0;
```

Comment:
I think we should experiment in the lab to find tighter bounds. 0 makes no sense (although it might do something reasonable), 300000 would likely take too long and our framerate would be too slow / blurred (assuming the driver actually used this which is unlikely).  I am worried it will be too dark / bright and the rate will keep drifting, then when lighting conditions return to normal and we try to return it could take a long time to converge back to the right values.

Reply:
sounds good! We do have a max_i, so it minimizes the recovery time